src/TrueMyth/Matcher.cs

```diff
@@ -0,0 +1,30 @@
+﻿using System;
+
+namespace TrueMyth
+{
+    /** A lightweight object defining how to handle each variant of a Maybe. */
+    public class Matcher<T>
+    {
+        IMaybe<T> _maybe;
+
+        public Matcher()
+        {
+            _maybe = Maybe.Nothing<T>();
+        }
+
+        internal Matcher(IMaybe<T> maybe) : this()
+            => _maybe = maybe;
+    }
+
+    public class Matcher<TIn, TOut>
+    {
```

Comment:
O.o - I don't even remember what this does

Reply:
The idea in the TS side from which this is ported is that you have a poor-man's pattern matching. E.g. in TS we might do:

```ts
function doSomethingWith(theMaybe: Maybe<string>) {
  theMaybe.match({
    Just: value => {
      console.log(value);
      return value.length;
    },
    Nothing: () => {
      console.log('nada');
      return 0;
    },
  });
}
```

This example doesn't show the utility, but it's handy when you have complicated things you need to do depending on which case you get. [This code does the same thing with `Result` and shows a bit more of the power.](https://github.com/ololabs/mobile-web-client/blob/develop/app/lib/customize/product.ts#L33:L40)

C♯'s pattern matching *might* be sophisticated enough to replace this; I haven't used it enough to know. If it *does*, then yeah, you could just drop this.

Reply:
In C♯ 7.0+, you can match on types like this:

```csharp
switch (maybe)
{
    case Just value:
        // stuff
    case Nothing:
        // more stuff
}
```

TBH, I'm not sure how this should actually work in C♯.  In F♯ it's clear how `Option` should work, because it greases the wheels for function composition - but you can't really compose functions in the same way in C♯, so I'm having a hard time imagining the use case there.

Reply:
Well, you *can*… you just end up writing a lot of `Func` types if you want to. 😭

I have no objection to dropping this in any case, especially for a "Just getting stuff up and running to use" start.

Reply:
Yeah.  I'm in favor of the "build it when we know we need it" approach.

Reply:
Same. At the time, I was basically just copying over the tests, then implementing everything to make things pass in the spirit of parity.